pkg/resources/management.cattle.io/v3/nodedriver/validator.go

```diff
@@ -0,0 +1,158 @@
+)
+
+var (
+	gvr = schema.GroupVersionResource{
+		Group:    "management.cattle.io",
+		Version:  "v3",
+		Resource: "nodedrivers",
+	}
+
+	driverInUse = &admissionv1.AdmissionResponse{
+		Result: &metav1.Status{
+			Status:  metav1.StatusFailure,
+			Message: "This driver is in use by existing nodes and cannot be disabled",
+		},
+		Allowed: false,
+	}
+)
+
+// Validator ValidatingWebhook for NodeDrivers
+type Validator struct {
+	admitter admitter
+}
+
+type admitter struct {
+	nodeCache controllersv3.NodeCache
+	dynamic   dynamicLister
+}
+
+// dynamicLister is an interface to abstract away how we list dynamic objects from k8s
+type dynamicLister interface {
+	List(gvk schema.GroupVersionKind, namespace string, selector labels.Selector) ([]runtime.Object, error)
+}
+
+// NewValidator returns a new Validator for NodeDriver resources
+func NewValidator(nodeCache controllersv3.NodeCache, dynamic *dynamic.Controller) admission.ValidatingAdmissionHandler {
+	return &Validator{admitter: admitter{
+		nodeCache: nodeCache,
+		dynamic:   dynamic,
+	}}
+}
+
+// GVR returns the GroupVersionKind for this CRD.
+func (v *Validator) GVR() schema.GroupVersionResource {
+	return gvr
+}
+
+// Operations returns list of operations handled by this validator.
+func (v *Validator) Operations() []admissionregistrationv1.OperationType {
+	return []admissionregistrationv1.OperationType{admissionregistrationv1.Update, admissionregistrationv1.Delete}
+}
+
+// ValidatingWebhook returns the ValidatingWebhook used for this CRD.
+func (v *Validator) ValidatingWebhook(clientConfig admissionregistrationv1.WebhookClientConfig) []admissionregistrationv1.ValidatingWebhook {
+	return []admissionregistrationv1.ValidatingWebhook{*admission.NewDefaultValidatingWebhook(v, clientConfig, admissionregistrationv1.ClusterScope, v.Operations())}
+}
+
+// Admitters returns the admitter objects used to validate machineconfigs.
+func (v *Validator) Admitters() []admission.Admitter {
+	return []admission.Admitter{&v.admitter}
+}
+
+// Admit is the entrypoint for the validator. Admit will return an error if it unable to process the request.
+// If this function is called without NewValidator(..) calls will panic.
+func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
+	oldObject, newObject, err := objectsv3.NodeDriverOldAndNewFromRequest(&request.AdmissionRequest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode object from request: %w", err)
+	}
+
+	// the check to see if the driver is being disabled is either when we're
+	// running a delete operation OR an update operation where the active flag
+	// toggles from true -> false
+	var disablingDriver bool
+	if (request.Operation == admissionv1.Delete && oldObject.Spec.Active) ||
+		(request.Operation == admissionv1.Update && !newObject.Spec.Active && oldObject.Spec.Active) {
+		disablingDriver = true
+	}
+
+	if !disablingDriver {
+		return admission.ResponseAllowed(), nil
+	}
```

Comment:
nit: could be simplified but this is more styltic so if you think it is easier to read the way you had it there is no need to change.
```suggestion
	if !(request.Operation == admissionv1.Delete && oldObject.Spec.Active) &&
		!(request.Operation == admissionv1.Update && !newObject.Spec.Active && oldObject.Spec.Active) {
		return admission.ResponseAllowed(), nil
	}
```